### PR TITLE
gcc: Add mold variant to use mold for linking

### PR DESCRIPTION
**Pull request by msimberg**
Triggered by https://github.com/spack/spack/pull/44034#issuecomment-2096707791.

This adds a `mold` variant to `gcc` which, when enabled, adds `mold` to GCC's search paths and `-fuse-ld=mold` to use `mold` without a user having to specify it themselves. With these changes I'm able to create an executable that reports:
```
String dump of section '.comment':
  [     0]  GCC: (SUSE Linux) 7.5.0
  [    18]  mold 2.31.0 (compatible with GNU ld)
  [    3d]  GCC: (Spack GCC) 13.2.0
```
It might make sense to add conflicts for when the `-fuse-ld=mold` flag doesn't work. The fallback option would be to use mold's `ld` alias in `libexec/mold` and only set `-B`. From what I can tell the current approach does not allow overriding the linker with a manually provided `-fuse-ld`, which may be a problem for some users but I don't know how much of a problem it is as long as the mold variant is off by default.

From mold's README:
> On Unix, the linker command (usually /usr/bin/ld) is indirectly invoked by the compiler driver (typically cc, gcc, or clang), which is in turn indirectly invoked by make or other build system commands.
> 
> If you can specify an additional command line option for your compiler driver by modifying the build system's config files, add one of the following flags to use mold instead of /usr/bin/ld:
> 
>     For Clang: pass -fuse-ld=mold
> 
>     For GCC 12.1.0 or later: pass -fuse-ld=mold
> 
>    For GCC before 12.1.0: the -fuse-ld option does not accept mold as a valid argument, so you need to use the -B option instead. The -B option tells GCC where to look for external commands like ld.
> 
>     If you have installed mold with make install, there should be a directory named /usr/libexec/mold (or /usr/local/libexec/mold, depending on your $PREFIX), and the ld command should be there. The ld is actually a symlink to mold. So, all you need is to pass -B/usr/libexec/mold (or -B/usr/local/libexec/mold) to GCC.
> 
> If you haven't installed ld.mold to any $PATH, you can still pass -fuse-ld=/absolute/path/to/mold to clang to use mold. However, GCC does not accept an absolute path as an argument for -fuse-ld.

Note: clang also supports `-fuse-ld=mold`, but I don't know how to enable that the same way in the `llvm` package, so I'm not going to deal with that (at least in this PR...).

### CODE REVIEW

**Comment by haampie** [comment]
:tada:

**Comment by haampie** [comment]
I need to check this bit:

```python
        # Binutils
        if spec.satisfies("+binutils"):
            binutils = spec["binutils"].prefix.bin
            options.extend(
                [
                    "--with-gnu-ld",
                    "--with-ld=" + binutils.ld,
                    "--with-gnu-as",
                    "--with-as=" + binutils.join("as"),
                ]
            )
```

which I believe is not entirely correct in that is uses absolute paths instead of file names, although now that I'm reading GCC's docs it does look sensible... (https://gcc.gnu.org/install/configure.html#with-as) In any case, it's likely we can delete those lines because we have this other `-B<prefix>` bit already.

Also I'm OK with `gcc +binutils +mold` like this PR allows, so that users can pick their favorite linker at runtime.

**Comment by msimberg** [comment]
> Also I'm OK with `gcc +binutils +mold` like this PR allows, so that users can pick their favorite linker at runtime.

Allowing `+binutils +mold` was an unintentional oversight by me :sweat_smile: though if you think it's ok to keep it like this that's fine by me. Note though that as far as I can tell the current version does not actually allow overriding the linker because (from a `--verbose` build) the `-fuse-ld=mold` and `-B` flags end up after flags provided on the command line.

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 1]
```diff
@@ -129,6 +129,7 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage, CompilerPackage):
         description="Compilers and runtime libraries to build",
     )
     variant("binutils", default=False, description="Build via binutils")
+    variant("mold", default=False, description="Use mold as the linker by default")
```
Comment (msimberg) [inline var/spack/repos/builtin/packages/gcc/package.py]: Commenting as a reminder to revisit before merging. Maybe add
```suggestion
    variant("mold", default=False, description="Use mold as the linker by default", when="@12:")
```
or
```suggestion
    variant("mold", default=False, description="Use mold as the linker by default")
    conflicts("+mold", when="@:11")
```

**Comment by haampie** [comment]
@alalazo knows a way to prepend.

Notice that +binutils also adds an assembler, and mold is only a linker, so we should be able to have both.

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 2]
```diff
@@ -973,6 +975,10 @@ def write_specs_file(self):
             # programs search path
             if self.spec.satisfies("+binutils"):
                 f.write(f"*self_spec:\n+ -B{self.spec['binutils'].prefix.bin}\n\n")
+
+            if self.spec.satisfies("+mold"):
+                f.write(f"*self_spec:\n+ -B{self.spec['mold'].prefix.bin} -fuse-ld=mold\n\n")
```
Comment (haampie) [inline var/spack/repos/builtin/packages/gcc/package.py]:
```suggestion
                f.write(f"*self_spec:\n+ -B{self.spec['mold'].prefix.bin} %{!fuse-ld*:-fuse-ld=mold}\n\n")
```

**Review by haampie — CHANGES_REQUESTED** [review]
The prepend trick `%rename x old_x` followed by `*x: <stuff> %(old_x)` isn't working for `self_spec`, but `%{!fuse-ld*:-fuse-ld=mold}` works to set defaults

**Comment by haampie** [comment]
Let's leave the `--with-ld=...` in, cause it's probably necessary for `gcc +bootstrap`'s own build. The `specs` file only applies after install.

Otherwise I think we can merge after some more testing :)

**Comment by haampie** [comment]
It still picks /usr/bin/ld on macOS, for some reason. Maybe because `--with-ld=...` disables search.

**Comment by haampie** [comment]
Hm, maybe `--with-ld` is interfering with GCC actually searching for a linker, cause it doesn't work on macOS:

```
$ gcc -print-prog-name=ld.mold
/path/to/mold-2.31.0-zedkgow2v6z74vxysvgswvziilmvaoqc/bin/ld.mold
```

but

**Review by haampie — APPROVED** [review]
Works fine on Linux, so I'm OK with this PR.

**Pull request merged**